core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/UnwrapCastInComparison.java

```diff
@@ -319,43 +326,48 @@ private Expression unwrapCast(ComparisonExpression expression)
+                    };
+                }
 
-            if (literalVsRoundtripped < 0) {
-                // cast rounded up
-                return switch (operator) {
-                    case EQUAL -> falseIfNotNull(cast.getExpression());
-                    case NOT_EQUAL -> trueIfNotNull(cast.getExpression());
-                    case IS_DISTINCT_FROM -> TRUE_LITERAL;
-                    case LESS_THAN, LESS_THAN_OR_EQUAL ->
-                        // We expect implicit coercions to be order-preserving, so the result of converting back from target -> source cannot produce a value
-                        // smaller than the next value in the source type
-                            new ComparisonExpression(LESS_THAN, cast.getExpression(), literalEncoder.toExpression(session, literalInSourceType, sourceType));
-                    case GREATER_THAN, GREATER_THAN_OR_EQUAL -> sourceRange.isPresent() && compare(sourceType, sourceRange.get().getMax(), literalInSourceType) == 0 ?
-                            new ComparisonExpression(EQUAL, cast.getExpression(), literalEncoder.toExpression(session, literalInSourceType, sourceType)) :
-                            new ComparisonExpression(GREATER_THAN_OR_EQUAL, cast.getExpression(), literalEncoder.toExpression(session, literalInSourceType, sourceType));
-                };
+                if (literalVsRoundtripped < 0) {
+                    // cast rounded up
+                    return switch (operator) {
+                        case EQUAL -> falseIfNotNull(cast.getExpression());
+                        case NOT_EQUAL -> trueIfNotNull(cast.getExpression());
+                        case IS_DISTINCT_FROM -> TRUE_LITERAL;
+                        case LESS_THAN, LESS_THAN_OR_EQUAL ->
+                                // We expect implicit coercions to be order-preserving, so the result of converting back from target -> source cannot produce a value
+                                // smaller than the next value in the source type
+                                new ComparisonExpression(LESS_THAN, cast.getExpression(), literalEncoder.toExpression(session, literalInSourceType, sourceType));
+                        case GREATER_THAN, GREATER_THAN_OR_EQUAL -> sourceRange.isPresent() && compare(sourceType, sourceRange.get().getMax(), literalInSourceType) == 0 ?
+                                new ComparisonExpression(EQUAL, cast.getExpression(), literalEncoder.toExpression(session, literalInSourceType, sourceType)) :
+                                new ComparisonExpression(GREATER_THAN_OR_EQUAL, cast.getExpression(), literalEncoder.toExpression(session, literalInSourceType, sourceType));
+                    };
+                }
+            }
+            else {
+                checkCondition(operator == EQUAL || operator == NOT_EQUAL || operator == IS_DISTINCT_FROM, INVALID_CAST_ARGUMENT, "Comparison other than equality found for unorderable type");
             }
```

Comment:
This is unnecessary, too. It's effectively validating that the ComparisonExpression that was passed to this method is valid, but that's an assumption we don't need to check.

Reply:
This is also a safeguard. In case there is a bug in parser/checker, it would be easier to understand what the cause is. And this also improves the code readability. But, if you strongly believe that this is not beneficial at all, I'm fine with deleting it.